### PR TITLE
Add a keybind to place items (implements and closes #13)

### LINKED ISSUE #13: Make configurable hotkey for placing items

### PR DESCRIPTION
**Pull request by 2008Choco**
This PR aims to add a configurable keybind to place items for both Forge and Fabric and resolve a longstanding issue/request, #13. Because Fabric's packet system isn't as refined as that of Forge's, it's slightly sloppy, but it gets the job done.

The keybind has its own category and its own language entries, therefore only the en_US.json language files have been updated because I'm not confident enough in my French and am unfamiliar with the other available languages. This is something that will need to be addressed in the future.

keybinding

By default, left shift is used as the keybinding but may be changed to any other key (as seen above). Left shift is used as the default to remain familiar to existing players while still giving the option of changing it if desired.

### CODE REVIEW

File: Forge/src/main/resources/assets/placeableitems/lang/en_us.json [hunk 1]
```diff
@@ -1,3 +1,6 @@
 {
-    "block.placeableitems.saddle_stand_block": "Saddle Stand"
+    "block.placeableitems.saddle_stand_block": "Saddle Stand",
+
+    "key.categories.placeableitems": "Placeable Items",
+    "key.placeableitems.place_special_item": "Place Special Item"
```
Comment (Ferdzz) [inline Forge/src/main/resources/assets/placeableitems/lang/en_us.json]: I think this should read "Place Item". Also if you have other things assigned to Shift (such as crouch), does it mark it as a conflict in the UI?

File: Forge/src/main/java/me/ferdz/placeableitems/event/ItemPlaceHandler.java [hunk 2]
```diff
@@ -2,25 +2,20 @@
...
 
+    private boolean holdingKey = false;
+
     @SubscribeEvent
     public void onRightClickBlock(final PlayerInteractEvent.RightClickBlock e) {
-        if (!e.getPlayer().isSneaking()) { // TODO: #13 Make configurable hotkey for placing items
-            return; // Abort if the user is not sneaking
+        if (!holdingKey) { // Abort if the user is not holding the keybind
+            return;
         }
```
Comment (Ferdzz) [inline Forge/src/main/java/me/ferdz/placeableitems/event/ItemPlaceHandler.java]: One of the things I've been wondering about this issue is that I don't know if the custom keybind should place the item directly on press of the keybind, or press of the keybind + right-click. I might have to playtest this. What do you think?

File: Forge/src/main/java/me/ferdz/placeableitems/PlaceableItems.java [hunk 3]
```diff
@@ -15,19 +15,31 @@
...
         FMLJavaModLoadingContext.get().getModEventBus().addListener(this::onCommonSetup);
-        DistExecutor.runWhenOn(Dist.CLIENT, () -> () -> FMLJavaModLoadingContext.get().getModEventBus().register(ClientListener.get()));
-        MinecraftForge.EVENT_BUS.register(new ItemPlaceHandler());
+        MinecraftForge.EVENT_BUS.register(this.placeHandler = new ItemPlaceHandler());
+
+        DistExecutor.runWhenOn(Dist.CLIENT, () -> () -> {
+            ClientListener clientListener = ClientListener.get();
+
+            FMLJavaModLoadingContext.get().getModEventBus().register(clientListener.getModBus());
+            MinecraftForge.EVENT_BUS.register(clientListener.getForgeBus());
```
Comment (Ferdzz) [inline Forge/src/main/java/me/ferdz/placeableitems/PlaceableItems.java]: This is starting to feel a bit messy. I would suggest moving the registering responsibility inside of ClientListener. Such as when .get() is called, the event buses are correctly registered without this constructor having to do it manually. Also this would allow you to get rid of the `getModBus` & `getForgeBus`

File: Forge/src/main/java/me/ferdz/placeableitems/client/ClientListener.java [hunk 4]
```diff
@@ -1,27 +1,79 @@
...
 import net.minecraftforge.fml.event.lifecycle.FMLClientSetupEvent;
 
+import org.lwjgl.glfw.GLFW;
+
 public final class ClientListener {
 
     private static final LazyLoadBase<ClientListener> INSTANCE = new LazyLoadBase<>(ClientListener::new);
 
-    private ClientListener() { }
+    private static final KeyBinding KEY_BINDING_PLACE_SPECIAL_ITEM = new KeyBinding("key." + PlaceableItems.MODID + ".place_special_item", GLFW.GLFW_KEY_LEFT_SHIFT, "key.categories." + PlaceableItems.MODID);
```
Comment (Ferdzz) [inline Forge/src/main/java/me/ferdz/placeableitems/client/ClientListener.java]: Why `SPECIAL`? I feel like `KEY_BINDING_PLACE_ITEM` is better. Could you go and rename them in the ids too (`place_special_item`)?

File: Forge/src/main/resources/assets/placeableitems/lang/en_us.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (2008Choco) [reply Forge/src/main/resources/assets/placeableitems/lang/en_us.json]: It would display the key. i.e. in French it would display `key.placeableitems.place_special_key`

Comment (2008Choco) [reply Forge/src/main/resources/assets/placeableitems/lang/en_us.json]: The reason I didn't go with `Place Item` is because Vanilla has one titled the same, though I guess if it's under its own category it shouldn't be much of an issue

File: Forge/src/main/java/me/ferdz/placeableitems/event/ItemPlaceHandler.java (hunk 2 above)
Comment (2008Choco) [reply Forge/src/main/java/me/ferdz/placeableitems/event/ItemPlaceHandler.java]: I found it most convenient to be holding the button and pressing right click. VeinMiner does something similar whereby you hold the tilde key and mine as you would normally. I think this is the least intrusive whereas having another keybind to place an item directly seems strange.

File: Forge/src/main/java/me/ferdz/placeableitems/PlaceableItems.java (hunk 3 above)
Comment (2008Choco) [reply Forge/src/main/java/me/ferdz/placeableitems/PlaceableItems.java]: I suppose that would work. It would also remove the need to have `@SubscribeEvent` annotations and public static methods.

**Comment by 2008Choco** [comment]
Another thing to note. I made changes to the protocol by adding a new packet. This under the assumption that a new version has not yet been released since the last PR. If one has been released, the protocol version will have to change. Otherwise, it's fine to remain the same.

File: Forge/src/main/resources/assets/placeableitems/lang/en_us.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (Ferdzz) [reply Forge/src/main/resources/assets/placeableitems/lang/en_us.json]: That's a shame, I hoped it would default to a locale that does have the translation. Not a huge problem

[the remark this answers is not in the document]
Comment (Ferdzz) [reply Forge/src/main/resources/assets/placeableitems/lang/en_us.json]: That's a shame, I hoped it would default to a locale that does have the translation. Not a huge problem

Comment (Ferdzz) [reply Forge/src/main/resources/assets/placeableitems/lang/en_us.json]: I think "Place Item" would be more desirable, even if the same name is used by MC

I'll do some tests tonight to see how the conflicts behave to make sure it's as intuitive as possible

Comment (2008Choco) [reply Forge/src/main/resources/assets/placeableitems/lang/en_us.json]: The only issue the conflicts have is red text for the keybinding in the control settings menu. Aside from that, it works flawlessly having it bound to the same key as any other. That aside, with the exception of Fabric. For some reason its keybind API is strange and `FabricKeyBinding.isPressed()` returns false when they're bound to the same key, though I worked around this by polling GLFW directly.

**TL;DR:** No gameplay difference. Works as expected.

Comment (2008Choco) [reply Forge/src/main/resources/assets/placeableitems/lang/en_us.json]: After consideration, perhaps we could append `(Hold)` to the keybind's name. `Place Item (Hold)`. This would at least be an indication of how it works

Comment (Ferdzz) [reply Forge/src/main/resources/assets/placeableitems/lang/en_us.json]: That's a good idea. I'll clone this locally and want to try it out before approving, see if I can get a text as clear as possible so users don't swarm me with the same questions ;)

File: Forge/src/main/java/me/ferdz/placeableitems/event/ItemPlaceHandler.java (hunk 2 above)
Comment (Ferdzz) [reply Forge/src/main/java/me/ferdz/placeableitems/event/ItemPlaceHandler.java]: You're right. After trying, this feels much more natural

File: Forge/src/main/resources/assets/placeableitems/lang/en_us.json (hunk 1 above)
Comment (Ferdzz) [reply Forge/src/main/resources/assets/placeableitems/lang/en_us.json]: Alright, after some testing, I think that adding "(Hold)" at the end may help users understand better.

It looks like there's a method inside `KeyBinding` called `conflicts`. Maybe it would be possible to override it and return false so it never appears red?

Comment (2008Choco) [reply Forge/src/main/resources/assets/placeableitems/lang/en_us.json]: As mentioned before, I'm just not sure that's the best approach. It seems slightly unsafe. Take for example, perhaps I have my tilde key assigned to both this keybind and that of inventory sorting. The configuration screen won't recognize it as a conflict and they'll function as intended, however I may accidentally organize my inventory while trying to place a block.

I could make that change and make it so it doesn't show it as conflicting, but in my personal opinion, I think it's better to show it as conflicting than not.

Alternatively, I could override `KeyBinding#conflicts()` and state that it won't conflict with the sneak key binding. I feel as though that may make the most sense.

Comment (Ferdzz) [reply Forge/src/main/resources/assets/placeableitems/lang/en_us.json]: That makes sense. Alright nevermind then

Comment (Ferdzz) [reply Forge/src/main/resources/assets/placeableitems/lang/en_us.json]: Would you mind adding the "(Hold)" at the end? Then we'll be good for final review

File: Fabric/src/main/java/me/ferdz/placeableitems/PlaceableItems.java [hunk 5]
```diff
@@ -1,20 +1,29 @@
...
 
     public static final String MODID = "placeableitems";
 
+    public static final Identifier PACKET_ID_NOTIFY_ITEM_PLACE_KEY = new Identifier(PlaceableItems.MODID, "notify_item_place_key");
+
     // Directly reference a log4j logger.
     public static Logger logger = Logger.getLogger("placeableitems");
 
+    private static boolean keyPressed = false;
+
```
Comment (Ferdzz) [inline Fabric/src/main/java/me/ferdz/placeableitems/PlaceableItems.java]: This probably doesn't belong in the mod class itself, it should be moved to something dedicated for the key presses / input handling

File: Forge/src/main/java/me/ferdz/placeableitems/network/CNotifySpecialItemPlaceKeyPacket.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+import net.minecraft.network.PacketBuffer;
+
+/**
+ * A client -> server packet to inform it of a client-sided keybind update for the
+ * special item place key.
+ *
+ * @author Parker Hawke - Choco
+ */
+public class CNotifySpecialItemPlaceKeyPacket {
```
Comment (Ferdzz) [inline Forge/src/main/java/me/ferdz/placeableitems/network/CNotifySpecialItemPlaceKeyPacket.java]: Should be renamed to `CNotifyItemPlaceKeyPacket `

File: Forge/src/main/java/me/ferdz/placeableitems/network/handler/AnonymousPacketHandler.java [hunk 7]
```diff
@@ -21,4 +22,12 @@
      */
     public void handleUpdateFluidHolder(SUpdateFluidHolderPacket packet, Supplier<NetworkEvent.Context> ctx);
 
+    /**
+     * Handle the {@link CNotifySpecialItemPlaceKeyPacket}.
+     *
+     * @param packet the packet to handle
+     * @param ctx the network context
+     */
+    public void handleNotifySpecialItemPlaceKey(CNotifySpecialItemPlaceKeyPacket packet, Supplier<NetworkEvent.Context> ctx);
```
Comment (Ferdzz) [inline Forge/src/main/java/me/ferdz/placeableitems/network/handler/AnonymousPacketHandler.java]: Same here, `Special` not needed

File: Fabric/src/main/java/me/ferdz/placeableitems/PlaceableItems.java (hunk 5 above)
Comment (2008Choco) [reply Fabric/src/main/java/me/ferdz/placeableitems/PlaceableItems.java]: Unfortunately I had no other place to put it. Having it be in the client class would cause class loader issues when accessed from the server and mixins will not load if its class contains static methods. Having this value be in its own class seems a bit awkward. I could still move it if you think it’s better in terms of organization but I figured I’d give my reasoning for putting it here instead

File: Forge/src/main/java/me/ferdz/placeableitems/network/CNotifySpecialItemPlaceKeyPacket.java (hunk 6 above)
Comment (2008Choco) [reply Forge/src/main/java/me/ferdz/placeableitems/network/CNotifySpecialItemPlaceKeyPacket.java]: Forgot about the packet. Whoops. Will rename that :)

**Comment by Ferdzz** [comment]
Thanks again for your help!
